### PR TITLE
Bugfix in HTTP Class + Enhance of LinkedIn Auth class

### PR DESCRIPTION
**Pull request by aWuttig**

### CODE REVIEW

File: lib/SimpleSAML/Utils/HTTP.php [hunk 1]
```diff
@@ -85,6 +85,10 @@ private static function getServerHTTPS()
             return false;
         }
 
+        if ($_SERVER['HTTPS'] === null) {
```
Comment (jaimeperez) [inline lib/SimpleSAML/Utils/HTTP.php]: What's exactly the bug here?

If `$_SERVER['HTTPS']` is null, there are two possible reasons for that:

* It is undefined, in which case, the previous code will get rid of it (`if (!array_key_exists('HTTPS', $_SERVER)) { return false; }`).
* It is explicitly defined to `null`, which is not something PHP is going to do itself, and therefore it must be someone, somewhere, doing it explicitly.

File: modules/authlinkedin/lib/Auth/Source/LinkedIn.php [hunk 2]
```diff
@@ -125,4 +120,23 @@ public function finalStep(&$state) {
 
 		$state['Attributes'] = $attributes;
 	}
+
+    /**
+     * @param $array
```
Comment (jaimeperez) [inline modules/authlinkedin/lib/Auth/Source/LinkedIn.php]: An actual description of what the method does, as well as the parameters and return values, would be helpful here.

File: modules/authlinkedin/lib/Auth/Source/LinkedIn.php [hunk 3]
```diff
@@ -125,4 +120,23 @@ public function finalStep(&$state) {
...
+     * @param string $prefix
+     *
+     * @return array
+     */
+    protected function flatten($array, $prefix = '') {
+        $result = [];
```
Comment (jaimeperez) [inline modules/authlinkedin/lib/Auth/Source/LinkedIn.php]: This is not valid syntax in PHP 5.3. It should be `$result = array();` instead.

File: modules/authlinkedin/lib/Auth/Source/LinkedIn.php [hunk 4]
```diff
@@ -125,4 +120,23 @@ public function finalStep(&$state) {
...
+        foreach($array as $key => $value) {
+            if(is_array($value)) {
+                $result = $result + $this->flatten($value, $prefix . $key . '.');
+            }
+            else {
+                $result[$prefix . $key] = [$value];
```
Comment (jaimeperez) [inline modules/authlinkedin/lib/Auth/Source/LinkedIn.php]: Same here:

```php
$result[$prefix.$key] = array($value);
```

File: modules/authlinkedin/lib/Auth/Source/LinkedIn.php [hunk 5]
```diff
@@ -125,4 +120,23 @@ public function finalStep(&$state) {
...
+
+    /**
+     * @param $array
+     * @param string $prefix
+     *
+     * @return array
+     */
+    protected function flatten($array, $prefix = '') {
+        $result = [];
+        foreach($array as $key => $value) {
```
Comment (jaimeperez) [inline modules/authlinkedin/lib/Auth/Source/LinkedIn.php]: According to the PSR-2 standard that we are following, there must be an space between the `foreach` keyword and the subsequent opening parenthesis.

File: modules/authlinkedin/lib/Auth/Source/LinkedIn.php [hunk 6]
```diff
@@ -125,4 +120,23 @@ public function finalStep(&$state) {
...
+            if(is_array($value)) {
```
Comment (jaimeperez) [inline modules/authlinkedin/lib/Auth/Source/LinkedIn.php]: Same here for the `if`:

```php
if (is_array($value)) {
```

File: modules/authlinkedin/lib/Auth/Source/LinkedIn.php [hunk 7]
```diff
@@ -125,4 +120,23 @@ public function finalStep(&$state) {
...
+                $result = $result + $this->flatten($value, $prefix . $key . '.');
+            }
+            else {
```
Comment (jaimeperez) [inline modules/authlinkedin/lib/Auth/Source/LinkedIn.php]: And in this case, the else should be surrounded by the braces, like this:

```php
} else {
```

File: lib/SimpleSAML/Utils/HTTP.php (hunk 1 above)
Comment (aWuttig) [inline lib/SimpleSAML/Utils/HTTP.php]: > It is explicitly defined to null, which is not something PHP is going to do itself, and therefore it must be someone, somewhere, doing it explicitly.
We use it together with shopware which is based on Zend Framework and maybe the framework set the HTTPS to null. I don`t know why. For us the simplest solution is to add this condition, because the key "HTTPS" exists but the value is null which means --> no HTTPS is activated.

File: modules/authlinkedin/lib/Auth/Source/LinkedIn.php (hunk 3 above)
Comment (aWuttig) [inline modules/authlinkedin/lib/Auth/Source/LinkedIn.php]: ah sorry :/ we are working with PHP7 ;-). You can change it if you want.

File: modules/authlinkedin/lib/Auth/Source/LinkedIn.php (hunk 5 above)
Comment (aWuttig) [inline modules/authlinkedin/lib/Auth/Source/LinkedIn.php]: ok.

**Comment by aWuttig** [comment]
In general we need this to authenticate users via linkedIn and get there location after this. The location is stored in an associative array by convention when getting this info from linkedIn response (see: https://developer.linkedin.com/docs/fields/location). 

The old method of putting the $userData into the $attributes array only iterates through the first level and this is not enough for this scenario.

Best Regards
André

File: lib/SimpleSAML/Utils/HTTP.php (hunk 1 above)
Comment (jaimeperez) [inline lib/SimpleSAML/Utils/HTTP.php]: Is it possible for you to trace this somehow to see who and why is setting this variable to `null`? It doesn't make much sense to me...

In any case, this should be a separate pull request so that we can keep things clean, and the actual fix is to call `empty()` in the return statement, so that any value evaluated by PHP as empty makes the method return `false`, conforming to the PHP manual:

```php
return !empty($_SERVER['HTTPS']);
```
